awsutil/awsutil_test.go

```diff
@@ -0,0 +1,76 @@
+package awsutil
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/stretchr/testify/require"
+)
+
+func TestECSTaskMeta(t *testing.T) {
+	taskRegion := "bogus-east-1"
+	nonTaskRegion := "some-other-region"
+	ecsMeta := ECSTaskMeta{
+		Cluster: "test",
+		TaskARN: fmt.Sprintf("arn:aws:ecs:%s:123456789:task/test/abcdef", taskRegion),
+		Family:  "task",
+	}
+
+	cases := []struct {
+		name         string
+		env          map[string]string
+		expectRegion string
+	}{
+		{
+			name:         "no-env",
+			env:          nil,
+			expectRegion: taskRegion,
+		},
+		{
+			name:         "aws-region",
+			env:          map[string]string{"AWS_REGION": nonTaskRegion},
+			expectRegion: nonTaskRegion,
+		},
+		{
+			name: "aws-default-region",
+			env: map[string]string{
+				// "AWS_DEFAULT_REGION is only read if AWS_SDK_LOAD_CONFIG is also set,
+				// and AWS_REGION is not also set."
+				"AWS_DEFAULT_REGION":  nonTaskRegion,
+				"AWS_SDK_LOAD_CONFIG": "1",
+			},
+			expectRegion: nonTaskRegion,
+		},
+		{
+			name: "aws-region-and-default-region",
+			env: map[string]string{
+				"AWS_REGION":          nonTaskRegion,
+				"AWS_DEFAULT_REGION":  "should-not-use-this-one",
+				"AWS_SDK_LOAD_CONFIG": "1",
+			},
+			expectRegion: nonTaskRegion,
+		},
```

Comment:
If the logic for the different region env vars lives in the AWS SDK, what is the reason for having the tests for it here? It seems like the only way to make them fail is to not call `session.NewSession()`, but then you wouldn't be creating an api client and the test would fail since the session would be nil. 

If we don't have any logic related to reading the different env variables, then we should remove tests for that since I think they're testing the behavior of the SDK rather than our own logic.

Reply:
That's fair. Though, it was only after writing these tests that I felt confident that we respect the `AWS_REGION`/`AWS_DEFAULT_REGION` environment variables if they are set... 

​While our code doesn't explicitly read those environment variables, it does configure the session and there are certainly ways to configure the session which would cause those those environment variables to be ignored (i.e. assume I'll write bad code in the future. Tests will save me.)

At minimum, I'd want to keep the first two test cases: (1) validate we _do_ fallback to the Task metadata with an empty env, and (2) validate we _don't_ cause `AWS_REGION` to be ignored if it's set.

Reply:
Agree on `#1`. For the `#2`, if there is a way we can write the code to ignore `AWS_REGION`, then it makes sense to test it. Ultimately, I'm advocating for tests that can fail and therefore can provide us feedback in the future. I couldn't come up with code that would make the `aws-region-and-default-region` test fail, for example, but maybe I just don't have vivid imagination 😄 

Either way, it's not blocking and so I'm OK with either decision!

Reply:
So for (2), if I accidentally invert a conditional in `awsutil.NewSession`:

```
        cfg := clientSession.Config
	if cfg.Region == nil || *cfg.Region != "" {  // bug!
		cfg.Region = aws.String(meta.Region())
	}
```

That results in us overwriting the environment-discovered region, despite not explicitly reading any of those variables. So including at least one test that sets AWS_REGION is useful, since it helps validate our implementation. 

But I agree we don't need multiple such tests with various combinations of environment variables.